core/unix/os.c

```diff
@@ -6191,6 +6191,7 @@ handle_close_pre(dcontext_t *dcontext)
 
     /* Xref PR 258731 - duplicate STDOUT/STDERR when app closes them so we (or
      * a client) can continue to use them for logging. */
+#ifdef STDFILE_FILENO
```

Comment:
This is important functionality for some use cases.  Many DR tools on UNIX, like similar tools such as `strace` or `valgrind`, provide their results by printing to the console, either stdout or stderr.  Without this dup-on-close, the output is lost when running apps like shells that like to close these files.  Compare:

```
$ bin64/drrun -stderr_mask 15 -c api/bin/libbbcount.so -- tcsh
<Starting application /bin/tcsh (66788)>
Client bbcount is running
<Stopping application /bin/tcsh (66789)>
Instrumentation results:
   1739101 basic block executions
      2071 basic blocks needed flag saving
      4287 basic blocks did not

% date
<Starting application /bin/date (66790)>
Client bbcount is running
Mon Apr  9 14:18:45 EDT 2018
<Stopping application /bin/date (66790)>
Instrumentation results:
     48939 basic block executions
       659 basic blocks needed flag saving
      2621 basic blocks did not

% exit
exit
<Stopping application /bin/tcsh (66788)>
Instrumentation results:
   6347122 basic block executions
      3473 basic blocks needed flag saving
      7672 basic blocks did not
```

```
$ bin64/drrun -stderr_mask 15 -no_dup_stdout_on_close -no_dup_stderr_on_close -c api/bin/libbbcount.so -- tcsh
<Starting application /bin/tcsh (66795)>
Client bbcount is running
<Stopping application /bin/tcsh (66796)>
Instrumentation results:
   1739099 basic block executions
      2071 basic blocks needed flag saving
      4285 basic blocks did not

% date
<Starting application /bin/date (66797)>
Client bbcount is running
Mon Apr  9 14:18:52 EDT 2018
% exit
exit
```

As you can see, without this feature the tool's results are just lost once in tcsh.

I see no reason to disable the key part of this feature: the dup of STDOUT and STDERR.  The only thing here that musl impacts is the update of the stdout/stderr used in the private libc copy, the privmod_std* update.  That is also less likely to affect the regular operation of many tools.

Also, rather than silently not working, could we have some kind of visible warning so that someone building this knows it's missing functionality that may cause existing tools to fail.

Reply:
Ah. Yeah, agreed. I didn't look at the code closely enough to notice that leaving this feature partially intact was a better option.

By "some kind of visible warning" do you mean a #warning, a `LOG(THREAD, ...` like I see nearby, or something else?

Reply:
> By "some kind of visible warning" do you mean a #warning, a LOG(THREAD, ... like I see nearby, or something else?

#warning and maybe also a `SYSLOG_INTERNAL_WARNING` which will both print to the log file and print to the console (in debug build only) (pre-close so should show up :)) so the user is aware something may be missing -- maybe that's too noisy but I would vote for that.